### DISCUSSION TITLE
Ego attention understanding

### DISCUSSION THREAD

**Discussion by kargarisaac** [eleurent/highway-env#177] [General]
Hi,

I'm interested to use your ego attention method in your paper with highway env. But I don't know about rl-agent and want to use it in my own PyTorch implementation. 
As I understood from your paper:

- Encoders in fig3 of the paper: first you get the state for the ego-agent and feed it into one MLP (embedding_layer) and the state of other agents into another MLP (others_embedding_layer). The output of this would be some vectors for each agent with the same size.

- Ego-attention (fig 3): Then you Lk, Lv, Lq that you learn them using GD to generate q0, k0,...,kn, v0,...,vn which are some scalar values. Then you create Q (scalar), K (vector), and V (vector) and use it in eq. 3. The output of this would be a weighted vector with a size similar to V. This is the output of ego-attention box in fig. 2. right?

- Decoder (fig.2): And this is an MLP net that generates action-value functions in DQN for example. 

Are these understandings correct? 

And another question: here N is the number of agents in the scene. Where do you handle variable number of agents?

**Comment by eleurent** [comment]
Hi @kargarisaac , thanks for your interest!

> Encoders in fig3 of the paper: first you get the state for the ego-agent and feed it into one MLP (embedding_layer) and the state of other agents into another MLP (others_embedding_layer). The output of this would be some vectors for each agent with the same size.

That's right! The same MLP could also be used for ego and other agents, I actually haven't really investigated if using different ones provides a big benefit.

> Ego-attention (fig 3): Then you Lk, Lv, Lq that you learn them using GD to generate q0, k0,...,kn, v0,...,vn which are some scalar values. Then you create Q (scalar), K (vector), and V (vector) and use it in eq. 3. The output of this would be a weighted vector with a size similar to V. This is the output of ego-attention box in fig. 2. right?

q0,k0,kn,v0,vn are not scalar but vectors of some dimension d_k (hyperparameter). Hence, Q, K and V are matrices obtained by stacking these vectors vertically (Q is a single row vector). Then yes, eq. 3 gives the output of the ego attention block.

> Decoder (fig.2): And this is an MLP net that generates action-value functions in DQN for example.

Absolutely. You can add standard tricks like e.g. a dueling architecture.

> And another question: here N is the number of agents in the scene. Where do you handle variable number of agents?

N is the number of other agents in one observation, and it can change from an observation to the next. It is naturally handled in the architecture since the same blocks Lq,Lr,Lv are used at each line (except 0), so you can simply use as many as the number of inputs you have. The aggregation of (3) is similar, it will juste have supplementary terms in the sum, but will always output a result of shape d_v, which does not depend on N. 

If you do manage to reproduce the results and release the code, could you kindly let me know?

**Comment by eleurent** [comment]
And of course, even if you do not use the rl-agents project, you can still take a look at [my own implementation](https://github.com/eleurent/rl-agents/blob/master/rl_agents/agents/common/models.py#L230)

**Comment by kargarisaac** [comment]
@eleurent Thank you for your quick answer. Actually, I couldn't find the code you linked. It's great to have it. Sure, I will let you know.

**Comment by kargarisaac** [comment]
I used your ego attention code and trained a model. here is the [code](https://github.com/kargarisaac/highway-env/blob/marl/scripts/dqn_ego_attention.py). The implementation is originally from cleanrl library.
and here is a video of that.

![hw](https://user-images.githubusercontent.com/24753756/92302398-8aac1680-ef74-11ea-8a6f-7f910a46dd02.gif)

**Comment by eleurent** [comment]
Thanks Isaac, that's really great! Do you see an improvement over a baseline (MLP) model?

**Comment by kargarisaac** [comment]
Yes, I see that the agent's understanding is more than before. It yields for others and then starts to move. I also see some crashes but I think it needs more training time. It was just a quick training on my laptop.
